client/proxmark3.c

```diff
@@ -70,16 +71,52 @@ byte_t* prx = rx;
 static void *uart_receiver(void *targ) {
 	struct receiver_arg *arg = (struct receiver_arg*)targ;
 	size_t rxlen;
+	bool need_reconnect = false;
 
 	while (arg->run) {
-		rxlen = 0;
-		if (uart_receive(sp, prx, sizeof(UsbCommand) - (prx-rx), &rxlen) && rxlen) {
-			prx += rxlen;
-			if (prx-rx < sizeof(UsbCommand)) {
+		if( need_reconnect ) {
+			sp = uart_open(sp_name);
```

Comment:
This will break on Android. Is there a better way to pass this from within the `main()` function instead?

See https://github.com/AndProx/AndProx/blob/b662c653d7ba33b570d71e34c949e3a8e7f17d8e/natives/src/main/cpp/natives.c#L82

Reply:
- For each of `uart_win32.c` and `uart_posix.c`, add a `char*` referring to the serial port path in their platform-specific `serial_port` structs.

- In `uart.h`, add a method `bool uart_reconnect(serial_port* sp)`.

- `uart_reconnect` can then be implemented in platform-specific ways, that takes in the serial port identifier stashed away earlier, and will setup the existing `serial_port` struct to work again. This returns true on success, or false on error.

- Whenever the application determines the need to reconnect, it calls `uart_reconnect`, which can operate in-place.  `uart_receiver` can have some retry logic.

Reply:
I have no idea what part of the above code will fail on Android, that link doesn't explain it.  The original PM3 code takes main()'s argv[1] and feeds it to uart_open().  sp_name is just a copy of argv[1].  In what cases will argv[1] not be a char* ?  If main() is changed to not feed argv[1] into uart_open then obviously the other instances of uart_open() will need to be changed too.  Surely replacing "char*" with whatever becomes appropriate at that time isn't that difficult...

The entire point of this patch is to simply detect a disconnection and attempt to reopen the same port.  It is not to rewrite the entire uart system.

Reply:
All of it will fail on Android, because the "serial port" is a reference to a bunch of JNI object IDs, which allow it to pass `uart_send` and `uart_receive` to Java method calls.

This is implemented in these files:

- https://github.com/AndProx/AndProx/blob/ae58e617cfd033f2d3ac94046a89291c790002a2/natives/src/main/cpp/natives.c#L122-L214
- https://github.com/AndProx/AndProx/blob/master/natives/src/main/cpp/uart_android.c
- https://github.com/AndProx/AndProx/blob/master/natives/src/main/java/au/id/micolous/andprox/natives/NativeSerialWrapper.java

PM3 doesn't get called by `main` either, so there is no `argv`.  The link I sent is one of the bootstrap functions which sets up PM3 similar to how `main` would.

This is done because there is no `/dev/ttyACM0` except on rooted devices where one has installed the `cdc_acm` kernel module.  Everything is passed around the [Android USB Host API](https://developer.android.com/guide/topics/connectivity/usb/host.html).  Requiring root is a major blocker to use on the platform.

Having an API call for "reconnect" means that it doesn't matter how `uart` is implemented, whether it's a string pointing to a device name, it's wrapped up in some other platform specific code, or it's part of some testing harness.

At least on Android I'd bring "reconnect" up into Java space, hit the USB stack on the head a few times, and then pass it back.

_As for precedent:_ @iceman1001 objected when I proposed removing `uart_{get,set}_speed` even though that's not actually used in mainline PM3, but was in his branch.

If that went in as-is, even when it was "internal", I would no longer be able to track upstream PM3 for Android until it was replaced with the API extensions I proposed.